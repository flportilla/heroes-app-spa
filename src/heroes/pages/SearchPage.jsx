import React from 'react'
import queryString from 'query-string'

import { useNavigate, useLocation } from 'react-router-dom'

import { useForm } from '../../hooks'
import { HeroCard } from '../components'
import { getHeroesByName } from '../helpers'

export const SearchPage = () => {

    const navigate = useNavigate();
    const { search } = useLocation();
    const { q = '' } = queryString.parse(search)

    const heroes = getHeroesByName(q);

    const { formState, onInputChange } = useForm({
        searchText: q
    });
    const { searchText } = formState

    const onSearchSubmit = (e) => {
        e.preventDefault();
        navigate(`?q=${searchText}`)
    }

    return (
        <>
            <h1>SearchPage</h1>
            <hr />

            <div className="row">

                <div className="col-5">
                    <h4>Searching</h4>
                    <hr />

                    <form onSubmit={onSearchSubmit}>
                        <input
                            type="text"
                            placeholder="Search a hero"
                            className="form-control"
                            name="searchText"
                            value={searchText}
                            onChange={onInputChange}
                            autoComplete="off"
                        />

                        <button
                            type="submit"
                            className="btn btn-outline-primary mt-1">
                            Search
                        </button>

                    </form>
                </div>

                <div className="col-7">

                    <h4>Results</h4>
                    <hr />
                    {
                        (q === '')
                            ? <div className="alert alert-info">
                                Search a hero
                            </div>
                            : (heroes.length === 0 && <div className="alert alert-danger">
                                No results for: <b>{q}</b>
                            </div>)
                    }

                    <>
                        {
                            heroes.map(hero =>
                                (<HeroCard key={hero.id} {...hero} />)
                            )
                        }
                    </>


                </div>
            </div>
        </>
    )
}
