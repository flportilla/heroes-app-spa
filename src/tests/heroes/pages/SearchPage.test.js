import { render, screen } from '@testing-library/react'
import { MemoryRouter } from 'react-router-dom'
import { SearchPage } from '../../../heroes/pages/SearchPage'

describe('Tests on <SearchPage /> component', () => {

    test('should show correctly with default values', () => {

        const { container } = render(
            <MemoryRouter>
                <SearchPage />
            </MemoryRouter>
        )

        expect(container).toMatchSnapshot()
    })

    test('should show Batman and the input with the value of queryString', () => {

        render(
            <MemoryRouter initialEntries={['/search?q=batman']}>
                <SearchPage />
            </MemoryRouter>
        )

        const input = screen.getByRole('textbox');
        expect(input.value).toBe('batman')

        const img = screen.getByRole('img')
        expect(img.src).toContain('assets/heroes/dc-batman.jpg')
    })



})