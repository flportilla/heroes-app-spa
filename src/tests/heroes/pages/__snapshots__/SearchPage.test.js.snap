// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests on <SearchPage /> component should show correctly with default values 1`] = `
<div>
  <h1>
    SearchPage
  </h1>
  <hr />
  <div
    class="row"
  >
    <div
      class="col-5"
    >
      <h4>
        Searching
      </h4>
      <hr />
      <form>
        <input
          autocomplete="off"
          class="form-control"
          name="searchText"
          placeholder="Search a hero"
          type="text"
          value=""
        />
        <button
          class="btn btn-outline-primary mt-1"
          type="submit"
        >
          Search
        </button>
      </form>
    </div>
    <div
      class="col-7"
    >
      <h4>
        Results
      </h4>
      <hr />
      <div
        class="alert alert-info"
      >
        Search a hero
      </div>
    </div>
  </div>
</div>
`;
